### PR TITLE
Update and improve the documentation

### PR DESCRIPTION
**Pull request by msm-code**
Currently the documentation is very poor and outdated.

Bring it up to date and improve.

### CODE REVIEW

File: docs/datasets.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# Datasets
+
+**TL;DR:** Every dataset represents a collection of indexed files. Datasets can be
+tagged and merged. Avoid too small datasets (<1000 files) or too many datasets (>100).
+
+Internally all files are indexed by ursadb and stored in datasets.
```
Comment (chivay) [inline docs/datasets.md:6]: `stored in datasets.` Are indexed files stored in datasets?

File: docs/datasets.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+Dataset is a collection of one or more indexes (read also about
+[index types](./indextypes.md)) It's the smallest collection of files
```
Comment (chivay) [inline docs/datasets.md:9]: `)) It's` missing dot (.) at the end of sentence.

File: docs/datasets.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+```
+docker-compoes exec ursadb ursacli -c 'dataset "dataset_id" taint "tag_name";'
+```
+
+You can also remove them in the same way:
+
+```
+ursacli
+ursadb> dataset "dataset_id" untaint "tag_name";
+```
```
Comment (chivay) [inline docs/datasets.md:58]: Maybe a note that `taint` here is not a bug in documentation?

File: docs/datasets.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+ursadb. Querying 1000 or 10000 datasets has much more overhead than querying 10 ones.
+
+In theory, you can merge all your datasets into one (using `compact all;` command
+repeatedly). This will work, but it's not recommended. That's because
+queries on really huge datasets can consume a lot of RAM (especially with
+wildcards), and because you can't benefit from streaming partial results.
+Empirically, 1M files per dataset is a good upper limit, but historically we've
+used datasets with many millions of samples and they work correctly.
+
+### tags
```
Comment (chivay) [inline docs/datasets.md:29]: Captitalize the section name

File: docs/datasets.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+```
+
+You can also remove them in the same way:
+
+```
+ursacli
+ursadb> dataset "dataset_id" untaint "tag_name";
+```
+
+### adding and removing files
```
Comment (chivay) [inline docs/datasets.md:60]: Capitalize section name

File: docs/indextypes.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
+# index types
```
Comment (chivay) [inline docs/indextypes.md:1]: Capitalize section name

File: docs/indextypes.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+ - `gram3` - classic trigrams
+ - `text4` - behave like 4grams for text, but can't query anything else
+ - `wide8` - behave like 4grams for utf16 ascii text, but can't query anything else
+ - `hash4` - 4grams packed into three bytes (with collisions)
+
+More indexes means more accurate results, but indexing will take longer, and more disk will be used.
+
+ - `gram3` is the most basic and important index, it should be used in all situations.
+ - `text4` and `wide8` indexes are wery useful for textual data. They improve
+    a lot of yara searches a lot and they don't need a lot of space, so they are
```
Comment (chivay) [inline docs/indextypes.md:19]: `a lot of yara searches a lot and they don't need a lot` :thinking:

File: docs/indextypes.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    almost always a good idea.
+ - `hash4` is the most complicated. It's not strictly necessary, and ursadb will work
+    without it smoothly. But if you have enough disk space and can afford extra
+    processing time, it'll reduce the number of false positives (and, in turn, make
+    ursadb faster).
+
+
+## details
```
Comment (chivay) [inline docs/indextypes.md:27]: Capitalize section name

File: docs/limits.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
+# limits
```
Comment (chivay) [inline docs/limits.md:1]: Capitalize section name

File: docs/ondiskformat.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
+# on-disk format
```
Comment (chivay) [inline docs/ondiskformat.md:1]: Captizalize

File: docs/ondiskformat.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+- [`dataset`](#dataset) file (json)
+- [`database`](#database) file (json)
+- [`index`](#index) file (binary)
+- [`names`](#names) file (plaintext)
+- [`namecache`](#namecache) file (binary)
+- [`itermeta`](#itermeta) file (json)
+- [`iterator`](#iterator) file (plaintext)
+
+## database
```
Comment (chivay) [inline docs/ondiskformat.md:13]: Capitalize

File: docs/ondiskformat.md [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    "iterators": {
+        "01ef4552": "itermeta.01ef4552.db.ursa",
+    },
+    "version": "1.3.2"
+}
+```
+
+## dataset
+
+Contains all necesary information about a [dataset](./datasets.md).
```
Comment (chivay) [inline docs/ondiskformat.md:38]: Not really enlightening. Maybe add some explanations about fields in the file?

File: docs/syntax.md [hunk 13]
```diff
@@ -0,0 +1,515 @@
...
+### Response format
+
+```
+{
+    "result": {
+        "status": "ok"
+    },
+    "type": "ok"
+}
+```
```
Comment (chivay) [inline docs/syntax.md:103]: What happens if there's an error?

File: docs/datasets.md (hunk 1 above)
Comment (msm-code) [reply docs/datasets.md]: good call, reworded a bit

File: docs/datasets.md (hunk 2 above)
Comment (msm-code) [reply docs/datasets.md]: :thumb

File: docs/datasets.md (hunk 3 above)
Comment (msm-code) [reply docs/datasets.md:59]: Why would it be a bug :(

Comment (msm-code) [reply docs/datasets.md:59]: (Repeated the word above, to show it's intentional)

File: docs/ondiskformat.md (hunk 10 above)
Comment (msm-code) [reply docs/ondiskformat.md]: section name

File: docs/ondiskformat.md (hunk 11 above)
Comment (msm-code) [reply docs/ondiskformat.md]: section name

File: docs/ondiskformat.md (hunk 12 above)
Comment (msm-code) [reply docs/ondiskformat.md:38]: Hard to say something interesting that's not also explained by the example. Still tried tho

File: docs/syntax.md (hunk 13 above)
Comment (msm-code) [reply docs/syntax.md]: I've added an explaination to the top of the document, with generic schemas for success and error results.

**Pull request merged**